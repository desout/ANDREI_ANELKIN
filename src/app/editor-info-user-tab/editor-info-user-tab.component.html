<div class="container">
<ng-container *ngIf="(userService.currentUser$ | async)?.name;then form; else loading"></ng-container>
<ng-template #form>

<form (ngSubmit)="onSubmit()" [formGroup]="userForm" #formDir="ngForm">
  <div class="form-group">
    <label>{{'INFO.USERNAME' | translate}}
      <input class="form-control" formControlName="name" [ngModel]="(userService.currentUser$ | async)?.name">
    </label>
    <div *ngIf="name.pending">{{'ERROR.VALIDATING' | translate}}</div>
    <div *ngIf="name.invalid && (name.dirty || name.touched);"
         class="alert alert-danger">
      <div *ngIf="name.errors['minlength']">
        {{'ERROR.MINLENGTH' | translate}} 4
      </div>
      <div *ngIf="name.errors['required']">
        {{'INFO.USERNAME' | translate}} {{'ERROR.REQUIRED' | translate}}
      </div>
      <div *ngIf="name.errors['suitName']">
        {{'ERROR.INCORRECTUSERNAME' | translate}}
      </div>
      <div *ngIf="name.errors['isBadName']">
        {{'ERROR.USERNOTAVAILABLE' | translate}}
      </div>
    </div>
  </div>

  <div class="form-group">

    <label>{{'INFO.BIRTHDAY' | translate}}
      <input class="form-control" formControlName="dateOfBirth" required [ngModel]="(userService.currentUser$ | async)?.dateOfBirth">
    </label>

    <div *ngIf="dateOfBirth.invalid && (dateOfBirth.dirty || dateOfBirth.touched)"
         class="alert alert-danger">
      <div *ngIf="dateOfBirth.errors['required']">
        {{'INFO.BIRTHDAY' | translate}} {{'ERROR.REQUIRED' | translate}}
      </div>
      <div *ngIf="dateOfBirth.errors['dateValue']">
        {{'ERROR.DATEFORMAT' | translate}} : 'YYYY/MM/DD', 'DD MMMM YYYY', 'DD-MMM-YY'
      </div>
    </div>
  </div>
  <div class="form-group">

    <label>{{'INFO.DATEOFLOGIN' | translate}}
      <input class="form-control" formControlName="dateOfFirstLogin" required [ngModel]="(userService.currentUser$ | async)?.dateOfFirstLogin">
    </label>

    <div *ngIf="dateOfFirstLogin.invalid && (dateOfFirstLogin.dirty || dateOfFirstLogin.touched)"
         class="alert alert-danger">
      <div *ngIf="dateOfFirstLogin.errors['required']">
        {{'INFO.DATEOFLOGIN' | translate}} {{'ERROR.REQUIRED' | translate}}
      </div>
      <div *ngIf="dateOfFirstLogin.errors['dateValue']">
        {{'ERROR.DATEFORMAT' | translate}}: 'YYYY/MM/DD', 'DD MMMM YYYY', 'DD-MMM-YY'
      </div>
    </div>
  </div>
  <div class="form-group">

    <label>{{'INFO.DATEOFNEXTNOTIF' | translate}}
      <input class="form-control" formControlName="dateNextNotification" required [ngModel]="(userService.currentUser$ | async)?.dateNextNotification">
    </label>

    <div *ngIf="dateNextNotification.invalid && (dateNextNotification.dirty || dateNextNotification.touched)"
         class="alert alert-danger">
      <div *ngIf="dateNextNotification.errors['required']">
        {{'INFO.DATEOFNEXTNOTIF' | translate}} {{'ERROR.REQUIRED' | translate}}
      </div>
      <div *ngIf="dateNextNotification.errors['dateValue']">
        {{'ERROR.DATEFORMAT' | translate}} : 'YYYY/MM/DD', 'DD MMMM YYYY', 'DD-MMM-YY'
      </div>
    </div>
  </div>
  <div class="form-group">

    <label>{{'INFO.INFORMATION' | translate}}
      <input class="form-control" formControlName="information" required [ngModel]="(userService.currentUser$ | async)?.information">
    </label>

    <div *ngIf="information.invalid && (information.dirty || information.touched)"
         class="alert alert-danger">
      <div *ngIf="information.errors['required']">
        {{'INFO.INFORMATION' | translate}} {{'ERROR.REQUIRED' | translate}}
      </div>
      <div *ngIf="information.errors['minlength']">
       {{'ERROR.MINLENGTH' | translate}} 4
      </div>
    </div>
  </div>
  <button type="submit" class="btn btn-default" [disabled]="userForm.invalid">{{'INFO.SUBMITBTN' | translate}}</button>
</form>
</ng-template>
<ng-template #loading><div class='lmask'></div></ng-template>
</div>
