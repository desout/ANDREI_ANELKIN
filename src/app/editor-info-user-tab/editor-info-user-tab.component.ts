import { Component, OnInit } from '@angular/core';
import {FormBuilder, FormGroup, Validators} from '@angular/forms';
import { UserService} from '../user.service';
import {TranslateService} from '@ngx-translate/core';
import {isNameTaken} from '../validators/user-name.validator';
import {DateValidator} from '../validators/date.validator';
import {startWith} from 'rxjs/operators';
import {LocalUser} from '../../../server/server/LocalUser';
import {isAvailableName} from '../validators/user-name-available.validator';

@Component({
  selector: 'app-editor-info-user-tab',
  templateUrl: './editor-info-user-tab.component.html',
  styleUrls: ['./editor-info-user-tab.component.scss']
})
export class EditorInfoUserTabComponent implements OnInit {
  user: LocalUser;
  userForm: FormGroup;

  getUserFromForm(): LocalUser {
    return {
      name:  this.userForm.get('name').value,
      dateOfBirth: this.userForm.get('dateOfBirth').value,
      dateOfFirstLogin: this.userForm.get('dateOfFirstLogin').value,
      dateNextNotification: this.userForm.get('dateNextNotification').value,
      information: this.userForm.get('information').value
    };
  }
  constructor(private userService: UserService, private formBuilder: FormBuilder, public translate: TranslateService) {}

  ngOnInit(): void {
    this.userForm = this.formBuilder.group({
      name: ['', Validators.compose([Validators.required, Validators.minLength(4)]),
        [isNameTaken.bind(this), isAvailableName(this.userService).bind(this)],
        'blur'],
      dateOfBirth: ['', Validators.compose([Validators.required, DateValidator])],
      dateOfFirstLogin: ['', Validators.compose([Validators.required, DateValidator])],
      dateNextNotification: ['', Validators.compose([Validators.required, DateValidator])],
      information: ['', Validators.compose([Validators.required, Validators.minLength(4)])]

    });

    this.userService.userUpdateHandle.pipe(startWith(this.user = this.userService.currentUser))
      .subscribe(( letUser ) => {
        this.user = letUser;
        if ( letUser) {
          this.userForm.patchValue( {
            name: this.user.name,
            dateOfBirth: this.user.dateOfBirth,
            dateOfFirstLogin: this.user.dateOfFirstLogin,
            dateNextNotification: this.user.dateNextNotification,
            information: this.user.information});
        }

      });
  }
  onSubmit() {
    const newUser: LocalUser = this.getUserFromForm();
    this.userService.updateUser(newUser).subscribe();
    this.userService.updateLocalUser(newUser);
  }

  get name() { return this.userForm.get('name'); }
  get dateOfBirth() { return this.userForm.get('dateOfBirth'); }
  get dateOfFirstLogin() { return this.userForm.get('dateOfFirstLogin'); }
  get dateNextNotification() { return this.userForm.get('dateNextNotification'); }
  get information() { return this.userForm.get('information'); }
}
