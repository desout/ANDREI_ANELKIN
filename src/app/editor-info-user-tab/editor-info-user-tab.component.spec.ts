import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { EditorInfoUserTabComponent } from './editor-info-user-tab.component';

describe('EditorInfoUserTabComponent', () => {
  let component: EditorInfoUserTabComponent;
  let fixture: ComponentFixture<EditorInfoUserTabComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [ EditorInfoUserTabComponent ]
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(EditorInfoUserTabComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });
});
