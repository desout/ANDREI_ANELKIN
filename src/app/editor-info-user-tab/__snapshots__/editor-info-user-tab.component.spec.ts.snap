// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorInfoUserTabComponent should create 1`] = `
<div
  id="root0"
  ng-version="7.0.3"
>
  <div
    class="container"
  >
    
    <div
      class="lmask"
    />
    
    
  </div>
</div>
`;
